### PR TITLE
Fix inconsistent results of params.deep_transform_keys

### LINKED ISSUE #50326: Inconsistent results of params.deep_transform_keys
### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
end

require "action_controller/railtie"
require "active_support"
require "active_support/core_ext/object/blank"
require "minitest/autorun"

class BugTest < Minitest::Test
  def test_stuff
    raw_parameters = { item: { exampleOne: 1 } }
    params = ActionController::Parameters.new(raw_parameters)
    #<ActionController::Parameters {"item"=>{"exampleOne"=>1}} permitted: false>

    expected = { 'item' => { 'example_one' => 1 } }
    assert_equal(expected, params.deep_transform_keys(&:underscore).to_unsafe_h) # true

    params[:item] # just access the item
    params
    #<ActionController::Parameters {"item"=>#<ActionController::Parameters {"exampleOne"=>1} permitted: false>} permitted: false>

    assert_equal(expected, params.deep_transform_keys(&:underscore).to_unsafe_h) # false
  end
end
```

### Expected behavior
`params.deep_transform_keys` should transform the keys of the params, including nested hashes, arrays, and `ActionController::Parameters`.
`params[key]` should not change the params object

### Actual behavior
`params.deep_transform_keys` fails to transform the keys of `ActionController::Parameters` nested inside params.
`params[key]` changes the params object.

### System configuration
**Rails version**: Rails 5+

**Ruby version**: Ruby 2.3+

### PR DESCRIPTION
**Pull request by iagopiimenta**
### Motivation / Background

Fix: #50326

### Detail

`params.deep_transform_keys` should transform the keys of the params, including nested hashes, arrays, and ActionController::Parameters objects.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by iagopiimenta** [comment]
Karma is failing to start browsers:

image

Do I need some special permission?

File: activesupport/lib/active_support/core_ext/hash/keys.rb [hunk 1]
```diff
@@ -134,6 +140,12 @@ def _deep_transform_keys_in_object!(object, &block)
           object[yield(key)] = _deep_transform_keys_in_object!(value, &block)
         end
         object
+      when defined?(ActionController::Parameters) && ActionController::Parameters
```
Comment (rafaelfranca) [inline activesupport/lib/active_support/core_ext/hash/keys.rb:143]: Active Support should not have any knowledge about `ActionController`. Can't this problem be solve by overriding those methods in `ActionController::Parameters`?

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

**Comment by cjavdev** [comment]
I think this change may have broken us (maybe we were depending on the wrong behavior here, but it feels like a pattern lots of users might use to convert from camel case to snake case). We convert from camel to snake case with deep_transform_keys, and before this change, it would give us the expected top level snake case key with the data nested, but after this change we now get an empty hash as the value because it seems to be overwritten instead of merged.


```rb
module Api
  module V1
    class WaitlistUsersController < Api::BaseController
      def create
        user = WaitlistUser.find_by(email: waitlist_user_params[:email].downcase)

        if user.nil?
          @waitlist_user = WaitlistUser.new(waitlist_user_params)
        else
          @waitlist_user = user
          user.update(waitlist_user_params)
        end

        if @waitlist_user.save
          render json: @waitlist_user, status: :created
        else
          render json: { errors: @waitlist_user.errors }, status: :unprocessable_entity
        end
      end

      private

      def waitlist_user_params
        params
          .deep_transform_keys!(&:underscore)
          .require(:waitlist_user)
          .permit(
            :email,
            :name,
          )
      end
    end
  end
end
```

Request:

```sh
POST /api/v1/waitlist_users

{
    "waitlistUser": { "email": "test@example.com" }
}
```
worked fine. 

Now:

```sh
bin/rails s -p 3001
=> Booting Puma
=> Rails 7.2.1 application starting in development
=> Run `bin/rails server --help` for more startup options


Started POST "/api/v1/waitlist_users" for ::1 at 2024-08-29 16:05:03 -0400
Processing by Api::V1::WaitlistUsersController#create as JSON
  Parameters: {"waitlistUser"=>{"email"=>"[FILTERED]"}, "waitlist_user"=>{}}
Completed 400 Bad Request in 1218ms (ActiveRecord: 23.2ms (0 queries, 0 cached) | GC: 212.9ms)

ActionController::ParameterMissing (param is missing or the value is empty: waitlist_user):

app/controllers/api/v1/waitlist_users_controller.rb:34:in `waitlist_user_params'
app/controllers/api/v1/waitlist_users_controller.rb:7:in `create'
app/controllers/concerns/set_locale.rb:13:in `set_locale'
```

**Comment by iagopiimenta** [comment]
I tested it, and it's working:
```ruby
irb(main):015> Rails.version
=> "7.2.1"
irb(main):016> params = ActionController::Parameters.new({ "waitlistUser": { "email": "TEST@example.com" } })
irb(main):017> waitlist_user_params = params.deep_transform_keys!(&:underscore).require(:waitlist_user).permit(:email, :name)
irb(main):018> waitlist_user_params[:email].downcase
=> "test@example.com"

```

@cjavdev make sure you don't have other things affecting your controller. You can try it with a fresh rails app.

**Comment by cjavdev** [comment]
Thanks @iagopiimenta I figured it out. It was because I had this enabled for `wrap_parameters`:

```rb
ActiveSupport.on_load(:action_controller) do
  wrap_parameters format: [:json]
end
```

This meant that the test for my scenario would be:

```rb
params = ActionController::Parameters.new({ "waitlistUser": { "email": "TEST@example.com" }, "waitlist_user": {} })
params.deep_transform_keys!(&:underscore).require(:waitlist_user).permit(:email, :name)[:email]
```

**Pull request merged**